### PR TITLE
feat: Migration of LCEVC DIL (Decoder Integration Layer) to LCEVC Dec (Decoder)

### PR DESCRIPTION
**Pull request by v-nova-vinod**
# Migration of LCEVC DIL (Decoder Integration layer) to LCEVC Dec (Decoder)

# Description
We are in the process of deprecating LCEVC DIL and migrating Integrations to LCEVC Dec. We also have a new library for the LCEVCDec on  [NPM](https://www.npmjs.com/package/lcevc_dec.js?activeTab=readme)

The new LCEVCDec has bug fixes for sync issues with VOD contents, and LCEVC Decode functionality for LIVE contents. There are new features to handle different streaming formats with timestamp offsets being passed to the Decoder.

The integration itself is very similar to the existing integration with some added parameters to existing functions and a new extern to handle additional features.

We have also added a quick start guide in the documentation and updated more LCEVC contents with different combination of containers and streaming formats.

# Documentation
All integration related documentation is available at [docs/design/lcevc-integration.md](https://github.com/v-novaltd/shaka-player/blob/lcevcDecJSMigration/docs/design/lcevc-integration.md)

Quick Start guide is available at [docs/tutorials/lcevc.md](https://github.com/v-novaltd/shaka-player/blob/lcevcDecJSMigration/docs/tutorials/lcevc.md)

# Executed Tasks
- [ ]  Signed Google CLA
- [ ]  Builds build/all.py successfully
- [ ]  Runs build/test.py successfully

### CODE REVIEW

File: docs/tutorials/lcevc.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+LCEVC Decoder library needs to be included in the HTML page: 
+ 
+- Local `npm install lcevc_dec.js` 
+```html 
+<script src="../node_modules/lcevc_dec.js/dist/lcevc_dec.min.js"></script> 
+``` 
+- Or use CDN.js for latest LCEVC Decoder libraries. 
+```html 
+<script src="https://unpkg.com/lcevc_dec.js@1.0.0/dist/lcevc_dec.min.js"></script> 
+```
```
Comment (joeyparrish) [inline docs/tutorials/lcevc.md:28]: unpkg.com != CDN.js

Please fix the prose above your sample code.

File: externs/lcevc.js [hunk 2]
```diff
@@ -1,25 +1,25 @@
...
 * compiler.
 *
 * @externs
 */
 
-// This empty namespace is declared to check if LcevcDil libraries are loaded.
+// This empty namespace is declared to check if LcevcDec libraries are loaded.
 var libDPIModule = {};
-var LcevcDil = {};
+var LCEVCdec = {};
```
Comment (joeyparrish) [inline externs/lcevc.js:10]: This is technically a breaking change AFAICT.  Apps that upgrade Shaka to a version with this PR will also have to be modified themselves to switch to the new module.  Unless you can provide some backward compatibility, this PR would require a major version bump in Shaka Player (`feat!:` in conventional commits syntax instead of `feat:`).

I'm not ready to merge this yet, as a major version bump triggers a year-long LTS release according to our maintenance policy.

Should we defer this?  Or would you like to try making it backward compatible?

File: docs/tutorials/lcevc.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+#### LCEVC setup via UI library 
+ 
+Sample setup using the Shaka Player UI library:  
+ 
+```html 
+<!DOCTYPE html> 
+<html> 
+  <head> 
+    <!-- Shaka Player UI compiled library and default CSS for player controls: --> 
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/shaka-player/4.3.5/shaka-player.ui.debug.min.js"></script>
```
Comment (joeyparrish) [inline docs/tutorials/lcevc.md:44]: This couldn't possibly work as written, since Shaka v4.3.5 doesn't support lcevc_dec.

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: docs/tutorials/lcevc.md (hunk 1 above)
Comment (v-nova-vinod) [reply docs/tutorials/lcevc.md]: Fixed this.

File: docs/tutorials/lcevc.md (hunk 3 above)
Comment (v-nova-vinod) [reply docs/tutorials/lcevc.md]: Moved this to use Shaka Player UI library generated in the dist locally.

File: externs/lcevc.js (hunk 2 above)
Comment (v-nova-vinod) [reply externs/lcevc.js:10]: The config parameters are not changed, hence the upgrade should not cause any issues in terms of setting up the player or the enhancement libraries. From a user perspective the only thing that would change is the library. Can you please provide more details about how we can make this backwards compatible? I am probably missing something.

File: demo/common/assets.js [hunk 4]
```diff
@@ -1322,23 +1322,74 @@ shakaAssets.testAssets = [
   // }}}
 
   // MPEG-5 LCEVC assets {{{
-  /* LCEVC Enabled Content with LCEVC Encoded Stream */
+  /* LCEVC Enhanced Contents */
```
Comment (avelad) [inline demo/common/assets.js:1325]: @v-nova-vinod In order to simplicate this PR, I moved the assets changes to another PR. https://github.com/shaka-project/shaka-player/pull/5522

Comment (avelad) [reply demo/common/assets.js:1325]: @v-nova-vinod can you rebase and remove the assets changes? Thanks

Comment (v-nova-vinod) [reply demo/common/assets.js]: Done

File: externs/lcevc.js (hunk 2 above)
Comment (joeyparrish) [reply externs/lcevc.js:10]: An application that includes the old DIL module will suddenly stop getting LCEVC support when this is merged.  That makes it a breaking change.

Comment (v-nova-vinod) [reply externs/lcevc.js:10]: Got it, so having functionality within the `lcevc_dec.js`  to add support to the old DIL module as the new Dec module in the `externs` would suffice right?
Can you please suggest if this is a valid approach to support the old library with the new implementation. Thanks!

Comment (joeyparrish) [reply externs/lcevc.js:10]: I'm sorry, I don't understand what you are proposing.  Let me try to explain backward compatibility in another way:

The standard for backward compatibility is that an application loading the old DIL module with a new Shaka Player should still work and get LCEVC support.  For Shaka Player to be backward compatible, it must continue to support the old thing.  Adding a new thing is fine, but removing an old thing is a breaking change.

Does that make sense?

Comment (v-nova-vinod) [reply externs/lcevc.js:10]: I have a latest commit addressing this issue, changes are made in the `lcevc_dec.js` to check for the availability of either `lcevc_dec.min.js` or `lcevc_dil.min.js`.  I have also added checks to the functions that are not supported. Please take a look and let me know if that addresses the issue. Thanks!

File: lib/lcevc/lcevc_dec.js [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    // This is a check to make the LCEVC Dec Integration is backwards compatible
+    // with previous Integration. This logic checks for both implementations
+    // and uses the one available.
+    if (typeof LCEVCdec !== 'undefined') {
+      this.lcevcLib_ = LCEVCdec;
+    } else {
+      if (typeof LcevcDil !== 'undefined') {
+        this.lcevcLib_ = LcevcDil;
+        this.toBeDeprecated_ = true;
+        shaka.log.alwaysWarn(
```
Comment (avelad) [inline lib/lcevc/lcevc_dec.js:135]: Include a `shaka.Deprecate.deprecateFeature`. See: https://github.com/shaka-project/shaka-player/blob/main/lib/player.js#L2099

And you can remove the alwaysWarn

**Review by avelad — CHANGES_REQUESTED** [review]
LGTM! Only a minor required change

File: lib/lcevc/lcevc_dec.js (hunk 5 above)
Comment (v-nova-vinod) [reply lib/lcevc/lcevc_dec.js]: Added the `deprecarteFeature` call for the case and removed `alwaysWarn`

**Comment by v-nova-vinod** [comment]
Hi @avelad @joeyparrish is this PR ready to be merged, or will you guys be checking and merging the PR on your side?

**Comment by avelad** [comment]
Hi @v-nova-vinod, for me it's ready to merge, but @joeyparrish 's review is still necessary... sorry!

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: lib/lcevc/lcevc_dec.js [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+  /**
+   * @param {HTMLVideoElement} media The video element that will be attached to
+   * LCEVC Decoder for input.
+   * @param {HTMLCanvasElement} canvas The canvas element that will be attached
+   * to LCEVC Decoder to render the enhanced frames.
+   * @param {shaka.extern.LcevcConfiguration} lcevcConfig LCEVC configuration
+   * object to initialize the LCEVC Decoder.
+   */
+  constructor(media, canvas, lcevcConfig) {
+    /** @private {?} */
```
[the remark this answers is not in the document]
Comment (v-nova-vinod) [reply lib/lcevc/lcevc_dec.js:30]: I had tried to put the type before, but for some reason JS compiler does not accept `LCEVCDec` as an acceptable type.

[the remark this answers is not in the document]
Comment (avelad) [reply lib/lcevc/lcevc_dec.js:30]: @joeyparrish ping!

[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/lcevc/lcevc_dec.js:30]: I will try to figure out the right syntax, push a commit to your fork if I find a fix, then approve and merge the PR

[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/lcevc/lcevc_dec.js:30]: I know the Closure compiler sucks and nobody knows how to work within its system, so I don't blame anybody for not figuring this out.  I'm ten-years deep in it by now, so I have seen some things.

I switched from `LCEVCdec` to `typeof LCEVCdec` (the interface to a module) and got the error `Missing type for typeof value. The value must be declared and const.`  So I went to your extern and marked the namespace with `/** @const */` like the others in that folder.

That got me to the place I said you would get, where you need a cast for `this.lcevcLib_.LcevcDil`.  I'll have a complete fix posted this afternoon (US/Pacific).

[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/lcevc/lcevc_dec.js:30]: `ERROR: Permission to v-novaltd/shaka-player.git denied to joeyparrish.`

@v-nova-vinod, when you send PRs, please consider using the `Allow edits from maintainers` feature to allow us to help you out.

In the meantime, here is the diff:

```diff
diff --git a/externs/lcevc.js b/externs/lcevc.js
index b884832ef..3134d3b87 100644
--- a/externs/lcevc.js
+++ b/externs/lcevc.js
@@ -6,7 +6,9 @@
 */
 
 // This empty namespace is declared to check if LcevcDec libraries are loaded.
+/** @const */
 var libDPIModule = {};
+/** @const */
 var LCEVCdec = {};
 
 /**
@@ -61,6 +63,13 @@ LCEVCdec.LCEVCdec = class {
   close() {}
 };
 
+/**
+ * The older module interface, for backward compatibility.
+ * Typed to the same interface, but under a different name.
+ * @type {typeof LCEVCdec.LCEVCdec}
+ */
+LCEVCdec.LcevcDil;
+
 /**
  * LCEVC Support Check
  */
diff --git a/lib/lcevc/lcevc_dec.js b/lib/lcevc/lcevc_dec.js
index ec4df6489..09aac2a8e 100644
--- a/lib/lcevc/lcevc_dec.js
+++ b/lib/lcevc/lcevc_dec.js
@@ -27,10 +27,12 @@ shaka.lcevc.Dec = class {
    * object to initialize the LCEVC Decoder.
    */
   constructor(media, canvas, lcevcConfig) {
-    /** @private {?} */
-    // LCEVC Decoder library based on availability, to check if either
-    // lcevc_dil or lcevc_dec is present.
-    this.lcevcLib_ = null;
+    /**
+     * LCEVC Decoder library based on availability, to check if either
+     * lcevc_dil or lcevc_dec is present.
+     * @private {typeof LCEVCdec}
+     */
+    this.lcevcLib_;
 
     /** @private {?LCEVCdec.LCEVCdec} */
     this.dec_ = null;
```

@v-nova-vinod, the reason this typing is important is because the compiler is able to catch lots of basic mistakes for us.  Turning it off for all calls into your module is a big gap in static analysis and compiler type safety.  Thanks for being patient and getting this fixed.  When these changes are applied, I would be happy to merge the PR.  Or @avelad can do it, since I'll be on vacation for the next week.

Thanks!

[the remark this answers is not in the document]
Comment (v-nova-vinod) [reply lib/lcevc/lcevc_dec.js]: Thank you for the explanation @joeyparrish , this is very helpful and thanks for the changes. I have made the changes and pushed. I will make sure for future PRs i will enable the option for the maintainer edits. Thanks again!!

[the remark this answers is not in the document]
Comment (avelad) [reply lib/lcevc/lcevc_dec.js]: @v-nova-vinod There are some errors:
```
ERROR: Unable to parse a tag's type expression for source file /home/runner/work/shaka-player/shaka-player/lib/lcevc/lcevc_dec.js in line 30 with tag title "private" and text "{typeof LCEVCdec}": Invalid type expression "typeof LCEVCdec": Expected "|" but "L" found.
ERROR: Unable to parse a tag's type expression for source file /home/runner/work/shaka-player/shaka-player/lib/lcevc/lcevc_dec.js in line 35 with tag title "private" and text "{typeof LCEVCdec}": Invalid type expression "typeof LCEVCdec": Expected "|" but "L" found.
```

[the remark this answers is not in the document]
Comment (v-nova-vinod) [reply lib/lcevc/lcevc_dec.js]: There seems to be some issue with the build script that is flagging a type issue.

[the remark this answers is not in the document]
Comment (v-nova-vinod) [reply lib/lcevc/lcevc_dec.js]: I think i know what might be happening here. With @joeyparrish 's fixes on the type as `typeof`,  the build process is successful but the when the docs are generated with `jsdoc`, it errors when it finds a space in the type name. 
I see that there is a custom version of `jsdoc` used so i tried updating the jsdoc version and it gave the error once and then there was no error seen. So we might need to take a call on how to address this issue.

[the remark this answers is not in the document]
Comment (v-nova-vinod) [reply lib/lcevc/lcevc_dec.js]: If this is run twice once the docs are generated when it errors, it does not flag this error. It is just the first time post some changes to the code. So i think it could be a false alarm from the docs generation part.

[the remark this answers is not in the document]
Comment (avelad) [reply lib/lcevc/lcevc_dec.js]: @joeyparrish ping!

[the remark this answers is not in the document]
Comment (v-nova-vinod) [reply lib/lcevc/lcevc_dec.js]: Hi @joeyparrish @avelad can you please suggest the next step on this? Thanks.

[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/lcevc/lcevc_dec.js]: Looking into it now.  Sorry for the delay.

[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/lcevc/lcevc_dec.js]: I am able to fix it by moving the `typeof` expression into a typedef in the externs.

However, permission was denied to me pushing to your fork because you are not allowing maintainer edits.

Please incorporate this diff:

```diff
diff --git a/externs/lcevc.js b/externs/lcevc.js
index 3134d3b87..c816df2a3 100644
--- a/externs/lcevc.js
+++ b/externs/lcevc.js
@@ -95,3 +95,11 @@ LCEVCdec.SupportObject.SupportStatus;
  * @type {string}
  */
 LCEVCdec.SupportObject.SupportError;
+
+/**
+ * Typedef for the module interface.  Both LCEVCdec (new module) and LcevcDil
+ * (old module) implement roughly the same interface.
+ *
+ * @typedef {typeof LCEVCdec}
+ */
+var LCEVCmodule;
diff --git a/lib/lcevc/lcevc_dec.js b/lib/lcevc/lcevc_dec.js
index 09aac2a8e..f28d3a324 100644
--- a/lib/lcevc/lcevc_dec.js
+++ b/lib/lcevc/lcevc_dec.js
@@ -30,7 +30,7 @@ shaka.lcevc.Dec = class {
     /**
      * LCEVC Decoder library based on availability, to check if either
      * lcevc_dil or lcevc_dec is present.
-     * @private {typeof LCEVCdec}
+     * @private {LCEVCmodule}
      */
     this.lcevcLib_;
 
```

[the remark this answers is not in the document]
Comment (v-nova-vinod) [reply lib/lcevc/lcevc_dec.js]: Thanks @joeyparrish!  That fixes it for me as well. I have made these changes and pushed.
I am not able to find the option to allow maintainers to allow commits on this PR. I think i will have to enable that when the PR is raised. I will keep this in mind in the next PR.

**Review by avelad — APPROVED** [review]

**Review by joeyparrish — APPROVED** [review]

**Comment by avelad** [comment]
Thanks @v-nova-vinod !
